### PR TITLE
Implementing Pareto distribution

### PR DESCRIPTION
**Pull request by vishwakftw**
This closes https://github.com/probtorch/pytorch/issues/62 , which is a mirror of https://github.com/pytorch/pytorch/issues/4434 .

I used implementation of `Cauchy` for reference, since `Chi2` just required calling `Gamma` with modified parameters.

### CODE REVIEW

**Comment by fritzo** [comment]
Wow, this is complex. It seems much simpler to implement this in pure python on top of PyTorch:
```py
class Pareto(Distribution):
    has_rsample = True
    def __init__(self, scale, alpha):
        self.scale, self.alpha = broadcast_all(scale, alpha)
        if isinstance(scale, Number) and isinstance(alpha, Number):
            batch_shape = torch.Size()
        else:
            batch_shape = self.scale.size()
        super(Pareto, self).__init__(batch_shape)
        self._exponential = Exponential(self.alpha)
    def rsample(self):
        return self.scale * torch.exp(self._exponential.rsample())
    def log_prob(self, value):
        # This probably has a little bug, but it should really only be 3 lines.
        exp_y = value / self.scale
        y = torch.log(exp_y)
        return self._exponential.log_prob(y) - torch.log(self.scale) - torch.log(exp_y)
```
What do you think?

**Comment by fritzo** [comment]
This PR should really only touch 4 files, the same 4 as `Chi2` but `torch/distributions/pareto.py` instead of `torch/distributions/cauchy.py`.

**Comment by vishwakftw** [comment]
@fritzo You are right. I thought implementing this on top of `Exponential` in Python could make it a bit less computational efficient and went with that. But I think your suggestion is actually better given that `Exponential` is implemented in C++/CUDA. I will revert the changes and add this one instead. Thank you for the suggestion :-)

File: torch/distributions/__init__.py [hunk 1]
```diff
@@ -34,14 +34,15 @@
 from .beta import Beta
 from .categorical import Categorical
 from .cauchy import Cauchy
+from .chi2 import Chi2
```
Comment (alicanb) [inline torch/distributions/__init__.py]: Don't do this. Chi2 imports Gamma, this will create circular import

Comment (vishwakftw) [reply torch/distributions/__init__.py]: I have changed it. Sorry about that.

File: torch/distributions/pareto.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        if isinstance(scale, Number) and isinstance(alpha, Number):
+            batch_shape = torch.Size()
+        else:
+            batch_shape = self.scale.size()
+        super(Pareto, self).__init__(batch_shape)
+
+    def rsample(self, sample_shape=torch.Size()):
+        shape = self._extended_shape(sample_shape)
+        exp_dist = self.alpha.new(shape).exponential_()
+        return self.scale.new(shape) * torch.exp(exp_dist)
```
Comment (fritzo) [inline torch/distributions/pareto.py]: I think this should be
```diff
- return self.scale.new(shape) * torch.exp(exp_dist)
+ return self.scale * torch.exp(exp_dist)
```
I believe `self.scale.new(shape)` returns a new tensor filled with uninitialized memory. The `new()` method is ok to use above since we immediately fill that uninitialized memory with `.exponential_()` values.

I'm concerned that the `test_pareto_sample()` tests did not catch this. Have you seen any local test failures?

File: test/test_distributions.py [hunk 3]
```diff
@@ -563,7 +577,65 @@ def test_gamma_sample_grad(self):
...
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    def test_pareto_sample(self):
+        set_rng_seed(0)  # see Note [Randomized statistical tests]
+        for scale, alpha in product([0.1, 1.0, 5.0], [0.1, 1.0, 10.0]):
+            self._check_sampler_sampler(Pareto(scale, alpha),
+                                        scipy.stats.pareto(alpha, scale=scale),
+                                        'Pareto(scale={}, alpha={})'.format(scale, alpha))
+
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    def test_pareto_sample_grad(self):
```
Comment (fritzo) [inline test/test_distributions.py]: This test is unnecessary. It is really only needed for `Gamma` and `Dirichlet` because they implement custom reparameterized gradients `standard_gamma_grad()` and `dirichlet_grad()`.

**Comment by vishwakftw** [comment]
@fritzo I have fixed them as per your suggestions.

**Comment by fritzo** [comment]
Looks like you're ready to close this PR and open up a PR against pytorch/pytorch. 
Feel free to add `math reviewed by @fritzo` in that PR description.

**Comment by fritzo** [comment]
To reiterate:

> I'm concerned that the test_pareto_sample() tests did not catch this. Have you seen any local test failures?

> Have you tested locally? Did you figure out why the test_pareto_sample() test did not catch the earlier bug?
